### ISSUE TITLE
Yield dependency doesn't allow to add a background task in its closing code

### ISSUE THREAD

**Issue by and-semakin** [tiangolo/fastapi#2866]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, Request, Depends, BackgroundTasks

app = FastAPI()


async def dep(background: BackgroundTasks):
    background.add_task(print, "before yield")
    yield "something"
    background.add_task(print, "after yield")

@app.get("/")
async def read_root(d: str = Depends(dep)):
    return {"Hello": "World", "dependency": d}
```

### Description

* Request the endpoint;
* The first background task (that is before yield) is added and executed;
* The second background task is silently ignored, there is no error and no warning.

### Environment

* OS: [e.g. Linux / Windows / macOS]: macOS
* FastAPI Version [e.g. 0.3.0]: `0.62.0`
* Python version: 3.8

### Additional context

I know that a task added after a response being sent has no chance to be executed. I guess there should be a warning or error if a user tries to do that, something like: `You can't add new background tasks after the yield keyword`. Also, probably, this behavior should be documented.

**Comment by falkben** [comment]
Agreed that it should at least be mentioned in the documentation. Would you be up for adding a note at the end of this section about this behavior and submitting a PR? https://fastapi.tiangolo.com/tutorial/background-tasks/#dependency-injection

**Comment by and-semakin** [comment]
Thanks for the answer @falkben! Sure, it needs to be documented. I'll send a PR soon.

Another solution could be extending Starlette's `BackgroundTask` and `BackgroundTasks` classes with a boolean that indicates if the tasks were executed or not. This way these classes will become one-time single-use objects. It won't be possible to add a new task to an already executed task collection or execute a task one more time. Need to discuss it with Starlette first.

**Comment by tiangolo** [comment]
Yep, dependencies with yield are closed *after* background tasks are done. This was to be able to use the same DB session.

But I see that in some cases it would make sense to close them *before* the response is sent. That's something I've been considering and thinking about how to properly implement.

All that is documented a bit here, wth a pretty diagram and everything: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#dependencies-with-yield-and-httpexception

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
